### PR TITLE
Flash Algorithm: Reworked documentation and template.

### PR DESCRIPTION
**Pull request by GuentherMartin**

### CODE REVIEW

File: Device/_Template_Flash/Target.ld [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    . = ALIGN(4);
+
+    *[Ff]lash[Dd]ev.o  ( .rodata .rodata.* ) ;
+
+    . = ALIGN(4);
+  } : DSCR
+}
+
+/* It seems that EXCLUDE_FILE can only be used once in the linker file.
+   All objects must be inside the (�) following the EXCLUDE_FILE directive.
```
Comment (JonatanAntoni) [inline Device/_Template_Flash/Target.ld:46]: There is a special char looking odd. Which character encoding is used for this file?

File: Device/_Template_Flash/FlashPrg.c [hunk 2]
```diff
@@ -50,18 +52,22 @@
...
-
 int Init (unsigned long adr, unsigned long clk, unsigned long fnc) {
 
   /* Add your Code */
-  return (0);                                  // Finished without Errors
+  UNUSED(adr);
+  UNUSED(clk);
+  UNUSED(fnc);
+  
+  return (0);                                            /* Finished without Errors */
```
Comment (JonatanAntoni) [inline Device/_Template_Flash/FlashPrg.c:70]: Why do we keep the return value in brackets?

File: Device/_Template_Flash/Abstract.txt [hunk 3]
```diff
@@ -4,28 +4,18 @@ Creating a new Algorithm
...
-- Open the dialog Project - Options for Target - Output and change the content
-  of the field Name of Executable to represent the device, for example
-  MyDevice.
+- Select an apropriate ARMCM# (ARMCM0 .. ARMCM7).
+- Choose a meaningful name for your project and the algorithm.
 - Adapt the programming algorithms in the file FlashPrg.c
 - Adapt the device parameters in the file FlashDev.c
-- Use Project - Build Target to generate the new Flash programming algorithm.
-  The output file (for example MyDevice.FLM) has to be added to the DFP.
+- Use VMSIS-Build to build the project.
```
Comment (JonatanAntoni) [inline Device/_Template_Flash/Abstract.txt:11]: Typo: I guess it shallbe CMSIS-Build

File: Device/_Template_Flash.deprecated/Abstract.txt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+Note
+----
+- Creating a Flash programming algorithm with MDK-Lite is not supported.
+- Flash programming algorithms use Read-Only Position Independent and
+  Read-Write Position Independent program code. These options are set in the
+  dialogs Project - Options for Target - C/C++ and
+  Project - Options for Target - Asm.
+- The dialog Project - Options for Target - Linker defines the linker scatter
+  file Target.lin. The error L6305 is disabled with �diag_suppress L6305.
```
Comment (JonatanAntoni) [inline Device/_Template_Flash.deprecated/Abstract.txt:28]: Special char with odd looking encoding. Should we use UTF-8 if special chars are required?

File: Device/_Template_Flash.deprecated/Abstract.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
+Creating a new Algorithm
```
Comment (JonatanAntoni) [inline Device/_Template_Flash.deprecated/Abstract.txt:1]: Do we need to keep this deprecated version? Is it still relevant? Or could we simply drop it from the head? One can always grab the old version from the history.

File: Device/_Example_Flash/Target.ld [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    . = ALIGN(4);
+
+    *[Ff]lash[Dd]ev.o  ( .rodata .rodata.* ) ;
+
+    . = ALIGN(4);
+  } : DSCR
+}
+
+/* It seems that EXCLUDE_FILE can only be used once in the linker file.
+   All objects must be inside the (�) following the EXCLUDE_FILE directive.
```
Comment (JonatanAntoni) [inline Device/_Example_Flash/Target.ld:46]: Special char.

File: Device/_Example_Flash/Abstract.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+Flash Programming Algorithm for STMicroelectronics STM32G4xx 512 KB Flash
```
Comment (JonatanAntoni) [inline Device/_Example_Flash/Abstract.txt:1]: The old abstract contained much more text. Did you reduce this by intent?

**Review by JonatanAntoni — COMMENTED** [review]
Looks good to me, with some minor remarks.

File: Device/_Template_Flash/Target.ld (hunk 1 above)
Comment (GuentherMartin) [reply Device/_Template_Flash/Target.ld:46]: I used a wrong encoding.
Anyhow, the comment will be deleted.

File: Device/_Example_Flash/Target.ld (hunk 6 above)
Comment (GuentherMartin) [reply Device/_Example_Flash/Target.ld:46]: I used a wrong encoding.
Anyhow, the comment will be deleted.

File: Device/_Template_Flash/FlashPrg.c (hunk 2 above)
Comment (GuentherMartin) [reply Device/_Template_Flash/FlashPrg.c:70]: I put brackets around the return values to make it more readable but I will remove the brackets.

File: Device/_Template_Flash/Abstract.txt (hunk 3 above)
Comment (GuentherMartin) [reply Device/_Template_Flash/Abstract.txt:11]: Yes, it is a typo.

File: Device/_Template_Flash.deprecated/Abstract.txt (hunk 4 above)
Comment (GuentherMartin) [reply Device/_Template_Flash.deprecated/Abstract.txt:28]: I will change the encoding to UTF-8.

File: Device/_Template_Flash.deprecated/Abstract.txt (hunk 5 above)
Comment (GuentherMartin) [reply Device/_Template_Flash.deprecated/Abstract.txt:1]: I would suggest to keep this version for consistency reasons.

File: Device/_Example_Flash/Abstract.txt (hunk 7 above)
Comment (GuentherMartin) [reply Device/_Example_Flash/Abstract.txt:1]: There was no 'old' abstract. This example did not exists before and the abstract describes what the example is about.

Comment (JonatanAntoni) [reply Device/_Example_Flash/Abstract.txt:1]: Ah, okay.. perhaps I messed example and template.

File: Device/_Template_Flash/FlashPrg.c (hunk 2 above)
Comment (JonatanAntoni) [reply Device/_Template_Flash/FlashPrg.c]: Well, we have this in many places. We can keep it if you think its more readable.

**Review by JonatanAntoni — APPROVED** [review]

**Comment by JonatanAntoni** [comment]
Can one of the admins verify this patch?

**Review by KeilChris — APPROVED** [review]
Looks good to me, ready to go.

**Pull request closed without merging**